llvm/lib/Object/SYCLBIN.cpp

```diff
@@ -0,0 +1,399 @@
+  StringRef RawModuleData{(*BinaryDataOrError)->getBufferStart(),
+                          (*BinaryDataOrError)->getBufferSize()};
+
+  IRModuleHeaderType IRMHeader{};
+  {
+    // Create metadata properties.
+    IRMHeader.MetadataOffset = MetadataByteTable.size();
+    llvm::util::PropertySetRegistry IRMMetadata;
+    // TODO: Determine state from the input.
+    IRMMetadata.add(llvm::util::PropertySetRegistry::SYCLBIN_IR_MODULE_METADATA,
+                    "type",
+                    /*SPIR-V*/ 0);
+    IRMMetadata.write(MetadataByteTableOS);
+    IRMHeader.MetadataSize =
+        MetadataByteTable.size() - IRMHeader.MetadataOffset;
+  }
+  IRMHeader.RawIRBytesOffset = BinaryByteTable.size();
+  IRMHeader.RawIRBytesSize = RawModuleData.size();
+  BinaryByteTableOS << RawModuleData;
+  return IRMHeader;
+}
+
+Expected<std::optional<SYCLBIN::NativeDeviceCodeImageHeaderType>>
+SYCLBIN::createNativeDeviceCodeImageHeader(
+    const SYCLBIN::ModuleDesc &Desc, const module_split::SplitModule &SM,
+    SmallString<0> &MetadataByteTable, raw_svector_ostream &MetadataByteTableOS,
+    SmallString<0> &BinaryByteTable, raw_svector_ostream &BinaryByteTableOS) {
+  // If there is no arch string, the module must be an IR module.
+  // Return with no native device code image in this case.
+  if (Desc.ArchString.empty())
+    return std::nullopt;
+
+  // Read the module data.
+  auto BinaryDataOrError =
+      llvm::MemoryBuffer::getFileOrSTDIN(SM.ModuleFilePath);
+  if (std::error_code EC = BinaryDataOrError.getError())
+    return createFileError(SM.ModuleFilePath, EC);
+  StringRef RawModuleData{(*BinaryDataOrError)->getBufferStart(),
+                          (*BinaryDataOrError)->getBufferSize()};
+
+  NativeDeviceCodeImageHeaderType NDCIHeader{};
+  {
+    // Create metadata properties.
+    NDCIHeader.MetadataOffset = MetadataByteTable.size();
+    llvm::util::PropertySetRegistry NDCIMetadata;
+    NDCIMetadata.add(llvm::util::PropertySetRegistry::
+                         SYCLBIN_NATIVE_DEVICE_CODE_IMAGE_METADATA,
+                     "arch", Desc.ArchString);
+    NDCIMetadata.write(MetadataByteTableOS);
+    NDCIHeader.MetadataSize =
+        MetadataByteTable.size() - NDCIHeader.MetadataOffset;
+  }
+  NDCIHeader.BinaryBytesOffset = BinaryByteTable.size();
+  NDCIHeader.BinaryBytesSize = RawModuleData.size();
+  BinaryByteTableOS << RawModuleData;
+  return NDCIHeader;
+}
+
+Expected<SmallString<0>>
+SYCLBIN::write(const SmallVector<SYCLBIN::ModuleDesc> &ModuleDescs) {
+  // TODO: Merge by properties, so overlap can live in the same abstract module.
+
+  FileHeaderType FileHeader;
+  FileHeader.Magic = MagicNumber;
+  FileHeader.Version = CurrentVersion;
+  FileHeader.AbstractModuleCount = 0;
+  FileHeader.IRModuleCount = 0;
+  FileHeader.NativeDeviceCodeImageCount = 0;
+
+  SmallString<0> MetadataByteTable;
+  raw_svector_ostream MetadataByteTableOS(MetadataByteTable);
+  SmallString<0> BinaryByteTable;
```

Comment:
Here is also a risk of hitting the upper boundary.

Reply:
See https://github.com/intel/llvm/pull/16873#discussion_r2046825822.